cni/network/invoker_cns.go

```diff
@@ -127,97 +132,167 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
-					{
-						Address: resultIPnet,
-						Gateway: ncgw,
-					},
-				},
-				Routes: []*cniTypes.Route{
-					{
-						Dst: network.Ipv4DefaultRouteDstPrefix,
-						GW:  ncgw,
-					},
-				},
+			ip, ncIPNet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
+			if ip == nil {
+				return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
 			}
-		} else if net.ParseIP(info.podIPAddress).To16() != nil {
-			addResult.ipv6Result = &cniTypesCurr.Result{
-				IPs: []*cniTypesCurr.IPConfig{
-					{
+
+			ncgw := net.ParseIP(info.ncGatewayIPAddress)
+			if ncgw == nil {
+				// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
+				if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) && (invoker.ipamMode != util.Overlay) {
+					return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+				}
+
+				if net.ParseIP(info.podIPAddress).To4() != nil { //nolint:gocritic
+					ncgw, err = getOverlayGateway(ncIPNet)
+					if err != nil {
+						return IPAMAddResult{}, err
+					}
+				} else if net.ParseIP(info.podIPAddress).To16() != nil {
+					ncgw = net.ParseIP(overlayGatewayV6IP)
+				} else {
+					return IPAMAddResult{}, errors.Wrap(err, "No podIPAddress is found: %w")
+				}
+			}
+
+			// construct ipnet for result
+			resultIPnet := net.IPNet{
+				IP:   ip,
+				Mask: ncIPNet.Mask,
+			}
+
+			if ip := net.ParseIP(info.podIPAddress); ip != nil {
+				if addResult.defaultCniResult.ipResult == nil {
+					addResult.defaultCniResult.ipResult = &cniTypesCurr.Result{}
+				}
+
+				defaultRouteDstPrefix := network.Ipv4DefaultRouteDstPrefix
+				if ip.To4() == nil {
+					defaultRouteDstPrefix = network.Ipv6DefaultRouteDstPrefix
+					addResult.ipv6Enabled = true
+				}
+
+				addResult.defaultCniResult.ipResult.IPs = append(addResult.defaultCniResult.ipResult.IPs,
+					&cniTypesCurr.IPConfig{
 						Address: resultIPnet,
 						Gateway: ncgw,
-					},
-				},
-				Routes: []*cniTypes.Route{
-					{
-						Dst: network.Ipv6DefaultRouteDstPrefix,
+					})
+				addResult.defaultCniResult.ipResult.Routes = append(addResult.defaultCniResult.ipResult.Routes,
+					&cniTypes.Route{
+						Dst: defaultRouteDstPrefix,
 						GW:  ncgw,
-					},
-				},
+					})
+				for _, route := range info.routes {
+					_, dst, routeErr := net.ParseCIDR(route.IPAddress)
+					if routeErr != nil {
+						return IPAMAddResult{}, fmt.Errorf("unable to parse destination %s: %w", route.IPAddress, routeErr)
+					}
+
+					gw := net.ParseIP(route.GatewayIPAddress)
+					if gw == nil {
+						return IPAMAddResult{}, fmt.Errorf("unable to parse gateway %s: %w", route.GatewayIPAddress, routeErr)
+					}
+
+					addResult.defaultCniResult.ipResult.Routes = append(addResult.defaultCniResult.ipResult.Routes,
+						&cniTypes.Route{
+							Dst: *dst,
```

Comment:
if code is same between primary and secondary, can we make it common?

Reply:
using common func getRoutes now